modules/dnn/src/layers/layer_norm.cpp

```diff
@@ -0,0 +1,172 @@
+            for (size_t i = 0; i < w_shape.size(); ++i)
+                CV_CheckEQ(w_shape[i], b_shape[i], "LayerNorm: bias dimensions does not match with weight dimensions");
+        }
+
+        // only one output is needed; Mean & InvStdDev are not needed
+        // in inference and should beomitted in onnx importer
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    class LayerNormInvoker : public ParallelLoopBody
+    {
+    public:
+        const Mat* src, *scale, *bias;
+        Mat *dst;
+
+        float epsilon;
+        int nstripes;
+
+        int total;
+        int stripeSize;
+        int normSize;
+
+        LayerNormInvoker() : src(0), scale(0), bias(0), dst(0), epsilon(0.f), nstripes(0), total(0), stripeSize(0), normSize(0) { }
+
+        static void run(const Mat* src, const Mat* scale, const Mat* b, Mat* dst, int axis, float epsilon, int nstripes)
+        {
+            CV_Assert(src->isContinuous());
+            CV_Assert(dst->isContinuous());
+            CV_CheckType(src->type(), CV_32F, "DNN/LayerNorm: only support float32");
+            CV_Assert(src->type() == dst->type());
+
+            LayerNormInvoker p;
+
+            p.src = src;
+            p.scale = scale;
+            p.bias = b;
+            p.dst = dst;
+
+            p.epsilon = epsilon;
+            p.nstripes = nstripes;
+
+            auto dstShape = shape(*dst);
+            p.total = std::accumulate(dstShape.begin(), dstShape.begin() + axis, 1, std::multiplies<int>());
+            p.stripeSize = (p.total + nstripes - 1) / nstripes;
+            p.normSize = std::accumulate(dstShape.begin() + axis, dstShape.end(), 1, std::multiplies<int>());
+
+            parallel_for_(Range(0, nstripes), p, nstripes);
+        }
+
+        void operator()(const Range& r) const CV_OVERRIDE
+        {
+            int stripeStart = r.start * stripeSize;
+            int stripeEnd = std::min(r.end * stripeSize, total);
+
+            float *dstData = (float *)dst->data;
+            const float *srcData = (const float *)src->data;
+            const float *scaleData = (const float *)scale->data;
+            const float *biasData = nullptr;
+            if (bias != nullptr)
+                biasData = (const float *)bias->data;
+
+            for (int ofs = stripeStart; ofs < stripeEnd; ++ofs)
+            {
+                const float* first = srcData + ofs * normSize;
+                float* dst_first = dstData + ofs * normSize;
+
+                float mean = 0;
+                float mean_square = 0;
+                for (int h = 0; h < normSize; ++h) {
+                    mean += first[h];
+                    mean_square += first[h] * first[h];
+                }
+                mean /= normSize;
+                mean_square = std::sqrt(mean_square / normSize - mean * mean + epsilon);
+                for (int h = 0; h < normSize; ++h) {
+                    if (biasData == nullptr) {
```

Comment:
Avoid "constant" `if` checks in for loops (there is "double" for loop).
This could be a template parameter.

Reply:
Could you give an example of template parameter? I googled it but did not find a related example. Or what about the following change:

```
if (biasData == nullptr) {
  // .. the double for loop
} else {
  // .. the double for loop with bias calculated
}
```

Reply:
It could be template bool haveBias, and then the code just changes from bias==nullptr to !haveBias. The compiler will optimize the check away. On the call site you add another if (bias==nulltr) func<false>() else func<true>(). The first thing that came to mind.

Reply:
> template bool haveBias

Thanks for the suggestion. From my perspective, the problem is actually how to eliminate duplicate code when the only difference is in the most inner loop.

Reply:
This approach solves that problem, doesn't it?

Reply:
> compiler can generate 2 optimal versions for us.

I start to get your point. Let me have a try tomorrow.

Reply:
With this change, layer perf managed to achieve the following performance:
```
[ RUN      ] Layer_LayerNorm.LayerNorm/0, where GetParam() = OCV/CPU
[ PERFSTAT ]    (samples=100   mean=0.02   median=0.02   min=0.01   stddev=0.01 (48.8%))
[       OK ] Layer_LayerNorm.LayerNorm/0 (3 ms)
```